File: app/src/main/java/com/rachelleignacio/listbucket/presentation/adapters/ListItemsAdapter.java
package com.rachelleignacio.listbucket.presentation.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.rachelleignacio.listbucket.R;
import com.rachelleignacio.listbucket.presentation.listeners.ListAdapterTouchListener;
import com.rachelleignacio.listbucket.domain.models.ListItem;
import com.rachelleignacio.listbucket.presentation.presenters.ListItemsFragmentPresenter;

import java.util.Collections;

/**
 * Created by rachelleignacio on 3/4/17.
 */

public class ListItemsAdapter extends RecyclerView.Adapter<ListItemsAdapter.ViewHolder>
        implements ListAdapterTouchListener {
    private java.util.List<ListItem> listItems;
    private ListItemsFragmentPresenter.View view;

    public ListItemsAdapter(java.util.List<ListItem> listItems, ListItemsFragmentPresenter.View view) {
        this.listItems = listItems;
        this.view = view;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.list_item_layout, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        ListItem listItem = listItems.get(position);
        holder.listItemNameTv.setText(listItem.getName());
    }

    @Override
    public int getItemCount() {
        return listItems.size();
    }

    @Override
    public void onRowDrag(int fromPosition, int toPosition) {
        if (fromPosition < toPosition) {
            for (int i = fromPosition; i < toPosition; i++) {
                Collections.swap(listItems, i, i+1);
            }
        } else {
            for (int i = fromPosition; i > toPosition; i--) {
                Collections.swap(listItems, i, i-1);
            }
        }
        notifyItemMoved(fromPosition, toPosition);
    }

    @Override
    public void onRowDismiss(int position) {
        view.onItemSwipedToDelete(listItems.get(position));
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public TextView listItemNameTv;

        public ViewHolder(View itemView) {
            super(itemView);
            listItemNameTv = (TextView) itemView.findViewById(R.id.item_name);
        }
    }
}

File: app/src/main/kotlin/com/rachelleignacio/listbucket/presentation/adapters/ListItemsAdapter.kt
package com.rachelleignacio.listbucket.presentation.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.rachelleignacio.listbucket.R
import com.rachelleignacio.listbucket.domain.models.ListItem
import com.rachelleignacio.listbucket.presentation.listeners.ListAdapterTouchListener
import com.rachelleignacio.listbucket.presentation.presenters.ListItemsFragmentPresenter

/**
 * Created by rachelleignacio on 8/29/17.
 */
class ListItemsAdapter(private val listItems: MutableList<ListItem>,
                       private val view: ListItemsFragmentPresenter.View)
    : RecyclerView.Adapter<ListItemsAdapter.ViewHolder>(), ListAdapterTouchListener {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
            LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.list_item_layout, parent, false)
                    .run { ViewHolder(this) }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.listItemNameTv.text = listItems[position].name
    }

    override fun getItemCount() = listItems.size

    override fun onRowDismiss(position: Int) { view.onItemSwipedToDelete(position) }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val listItemNameTv: TextView = itemView.findViewById(R.id.item_name)
    }
}